src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -1838,42 +1892,68 @@ internal static object[] CreateAttributeArrayHelper(RuntimeType caType, int elem
         }
         #endregion
 
-        #region Private Static FCalls
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern void _ParseAttributeUsageAttribute(
-            IntPtr pCa, int cCa, out int targets, out bool inherited, out bool allowMultiple);
-        private static void ParseAttributeUsageAttribute(
-            ConstArray ca, out AttributeTargets targets, out bool inherited, out bool allowMultiple)
-        {
-            _ParseAttributeUsageAttribute(ca.Signature, ca.Length, out int _targets, out inherited, out allowMultiple);
-            targets = (AttributeTargets)_targets;
-        }
-
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern object _CreateCaObject(RuntimeModule pModule, RuntimeType type, IRuntimeMethodInfo pCtor, byte** ppBlob, byte* pEndBlob, int* pcNamedArgs);
-        private static object CreateCaObject(RuntimeModule module, RuntimeType type, IRuntimeMethodInfo ctor, ref IntPtr blob, IntPtr blobEnd, out int namedArgs)
-        {
-            byte* pBlob = (byte*)blob;
-            byte* pBlobEnd = (byte*)blobEnd;
-            int cNamedArgs;
-            object ca = _CreateCaObject(module, type, ctor, &pBlob, pBlobEnd, &cNamedArgs);
-            blob = (IntPtr)pBlob;
-            namedArgs = cNamedArgs;
-            return ca;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "CustomAttribute_CreateCustomAttributeInstance")]
+        private static partial void CreateCustomAttributeInstance(
+            QCallModule pModule,
+            ObjectHandleOnStack type,
+            ObjectHandleOnStack pCtor,
+            ref IntPtr ppBlob,
+            IntPtr pEndBlob,
+            out int pcNamedArgs,
+            ObjectHandleOnStack instance);
+
+        private static object CreateCustomAttributeInstance(RuntimeModule module, RuntimeType type, IRuntimeMethodInfo ctor, ref IntPtr blob, IntPtr blobEnd, out int namedArgs)
+        {
+            if (module is null)
+            {
+                throw new ArgumentNullException(SR.Arg_InvalidHandle);
+            }
+
+            object? result = null;
+            CreateCustomAttributeInstance(
+                new QCallModule(ref module),
+                ObjectHandleOnStack.Create(ref type),
+                ObjectHandleOnStack.Create(ref ctor),
+                ref blob,
+                blobEnd,
+                out namedArgs,
+                ObjectHandleOnStack.Create(ref result));
+            return result!;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "CustomAttribute_CreatePropertyOrFieldData", StringMarshalling = StringMarshalling.Utf16)]
+        private static partial void CreatePropertyOrFieldData(
+            QCallModule pModule,
+            ref IntPtr ppBlobStart,
+            IntPtr pBlobEnd,
+            StringHandleOnStack name,
+            [MarshalAs(UnmanagedType.U1)] out bool bIsProperty,
```

Comment:
Do we have any unified preference for using U1 or I4 for bool in QCall? Both options should work.

Reply:
https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/corelib.md#qcall-functional-behavior recommends `BOOL`

Reply:
> Both options should work.

This is not true. The type on the native side is `bool`, which @jkotas is correct it should `BOOL`. If the native type were `BOOL`, then yes either U1 or U4 would work in theory.

Reply:
> The type on the native side is `bool`

I mean that other QCALLs use `BOOL` at unmanaged side.